### PR TITLE
Validate that incompatible advanced materials don't intersect

### PR DESCRIPTION
**Pull request by caseyflex**
Certain advanced materials (custom, nonlinear, time modulated, fully anistropic) are not allowed to intersect, but this was not validated on the frontend.

@tylerflex test_IO fails because the old tests/sims/simulation_* files are based on utils.py that actually had such materials intersecting. I revised utils.py to fix this, but test_IO still fails on the old sim files -- how do I address this?

### CODE REVIEW

**Comment by tylerflex** [comment]
> @tylerflex test_IO fails because the old tests/sims/simulation_* files are based on utils.py that actually had such materials intersecting. I revised utils.py to fix this, but test_IO still fails on the old sim files -- how do I address this?

Hm, might need to try manually fixing the oldest sim file? I'm not sure I see another way around it. 

If we want to escape the error, we could try calling `pytest.raises` and passing either a `match` or `escape` for that specific validation message, but I think it would be better to just fix the file if it hopefully isn't too much of a pain. Looks like it's just adjusting the size/center of a couple structures?

**Comment by caseyflex** [comment]
> > @tylerflex test_IO fails because the old tests/sims/simulation_* files are based on utils.py that actually had such materials intersecting. I revised utils.py to fix this, but test_IO still fails on the old sim files -- how do I address this?
> 
> Hm, might need to try manually fixing the oldest sim file? I'm not sure I see another way around it.
> 
> If we want to escape the error, we could try calling `pytest.raises` and passing either a `match` or `escape` for that specific validation message, but I think it would be better to just fix the file if it hopefully isn't too much of a pain. Looks like it's just adjusting the size/center of a couple structures?

yeah, I can do - it's all the old sim files from 2.4 - 2.6

File: tidy3d/components/simulation.py
Comment (tylerflex) [inline tidy3d/components/simulation.py:934]: I think just `if type1 == type2` should be fine

File: tidy3d/components/simulation.py
Comment (tylerflex) [inline tidy3d/components/simulation.py:927]: probably easier to just iterate over
```py
for ind1, type1 in zip(advanced_indices, advanced_types):
```
and same below

**Comment by tylerflex** [comment]
> yeah, I can do - it's all the old sim files from 2.4 - 2.6
I think if you delete all but 2.4 and re-run the tests, it should auto generate the other ones. but maybe try in a backup branch so it doesn't mess things up..

**Comment by caseyflex** [comment]
> > yeah, I can do - it's all the old sim files from 2.4 - 2.6
> 
> I think if you delete all but 2.4 and re-run the tests, it should auto generate the other ones. but maybe try in a backup branch so it doesn't mess things up..

Thanks I'll try that -- should be better, I could fix the jsons by hand but not sure how to manually edit the h5 files.

**Comment by caseyflex** [comment]
> > > yeah, I can do - it's all the old sim files from 2.4 - 2.6
> > 
> > 
> > I think if you delete all but 2.4 and re-run the tests, it should auto generate the other ones. but maybe try in a backup branch so it doesn't mess things up..
> 
> Thanks I'll try that -- should be better, I could fix the jsons by hand but not sure how to manually edit the h5 files.

It didn't auto-generate the old ones -- it just auto-generated 2.6.0.h5 and 2.6.0.json. The tests then passed, but the old sim h5 and json files are gone.

**Review by tylerflex — CHANGES_REQUESTED** [review]
So this might be a good time to revisit this question we had a few months ago about how to best define incompatibilities between different parts of the spec. While I think the code here is ok, I'm wondering if we can define the incompatibilities a bit more as a data structure or property/method so we can generalize this in the future. Some idea:

```py
class AbstractMedium:

    @property
    def _is_advanced(self) -> bool:
        """Whether a given medium is 'advanced' <- might need a more specific word for it."""

    def _compatible_with(other: AbstractMediumMedium) -> bool:
        """Whether these two mediums are compatible if in structures that intersect."""
```

```py
class Structure:

    def _compatible_with(other: Structure) -> bool:
        """Whether these two structures are compatible."""
        if not (self.medium._is_advanced and other.medium.is_advanced):
            return True
        
       # note, if 1st condition fails, the 2nd won't get triggered.
        if not self.medium._is_compatible(other.medium) and self.intersects(other):
            return False
```

And then I think the validator becomes much simpler to implement?

```py
def _validate_advanced_material_intersections(cls, val, values):
    for i, structure_i in enumerate(structures):
        for j, structure_j in enumerate(structures[i+1:]:
             if not structure_i.is_compatible(structure_j):
                 raise ...
```

What do you think?

**Comment by tylerflex** [comment]
> It didn't auto-generate the old ones -- it just auto-generated 2.6.0.h5 and 2.6.0.json. The tests then passed, but the old sim h5 and json files are gone.

Shoot.. hopefully there's a way to find and replace all of these otherwise I don't know exactly how else to handle it. (I think we need the intermediate files for the updater tests)

File: tidy3d/components/simulation.py
Comment (caseyflex) [reply tidy3d/components/simulation.py:927]: I think the reason I did it this way is to start the inner for loop at j = i+1, so I don't check each pair twice

**Comment by tylerflex** [comment]
> Thanks I'll try that -- should be better, I could fix the jsons by hand but not sure how to manually edit the h5 files.

hm, I was going to say to load a simulation from json and then export an h5 `to_file` but that would require checking out each version and sounds like a huge pain too.

Just FYI the hdf5 file has a dataset at the top level called "JSON_STR" or something which just contains the json string. Maybe that could be directly edited as a last resort.

File: tidy3d/components/simulation.py
Comment (caseyflex) [reply tidy3d/components/simulation.py:927]: If we go with your suggestion for adding the medium.is_compatible(other_medium) function, then that breaks the symmetry a bit and we might end up checking each pair twice. Which is ok, just some redundancy which could actually help prevent errors

**Comment by caseyflex** [comment]
> > Thanks I'll try that -- should be better, I could fix the jsons by hand but not sure how to manually edit the h5 files.
> 
> hm, I was going to say to load a simulation from json and then export an h5 `to_file` but that would require checking out each version and sounds like a huge pain too.
> 
> Just FYI the hdf5 file has a dataset at the top level called "JSON_STR" or something which just contains the json string. Maybe that could be directly edited as a last resort.

yeah both a bit unfortunate, but I can try...

File: tidy3d/components/simulation.py
Comment (tylerflex) [reply tidy3d/components/simulation.py:927]: yea, in my code snippet for the new suggestion, I also start the loop at i+1 in the validator , so that should resolve it?

Comment (caseyflex) [reply tidy3d/components/simulation.py:927]: ah, right.

**Comment by caseyflex** [comment]
> So this might be a good time to revisit this question we had a few months ago about how to best define incompatibilities between different parts of the spec. While I think the code here is ok, I'm wondering if we can define the incompatibilities a bit more as a data structure or property/method so we can generalize this in the future. Some idea:
> 
> ```python
> class AbstractMedium:
> 
>     @property
>     def _is_advanced(self) -> bool:
>         """Whether a given medium is 'advanced' <- might need a more specific word for it."""
> 
>     def _compatible_with(other: AbstractMediumMedium) -> bool:
>         """Whether these two mediums are compatible if in structures that intersect."""
> ```
> 
> ```python
> class Structure:
> 
>     def _compatible_with(other: Structure) -> bool:
>         """Whether these two structures are compatible."""
>         if not (self.medium._is_advanced and other.medium.is_advanced):
>             return True
> 
>        if self.medium.type == other.medium.type:
>             return True
>         
>        # note, if 1st condition fails, the 2nd won't get triggered.
>         if not self.medium._is_compatible(other.medium) and self.intersects(other):
>             return False
> ```
> 
> And then I think the validator becomes much simpler to implement?
> 
> ```python
> def _validate_advanced_material_intersections(cls, val, values):
>     for i, structure_i in enumerate(structures):
>         for j, structure_j in enumerate(structures[i+1:]:
>              if not structure_i.is_compatible(structure_j):
>                  raise ...
> ```
> 
> What do you think?

I restructured according to your suggestion and I think it made the logic much cleaner. I also got rid of the notion of the "type" of an advanced medium, since in principle it could have multiple.

File: tidy3d/components/medium.py
Comment (tylerflex) [inline tidy3d/components/medium.py:671]: I'd prefer `is_nonlinear`, `is_custom`, etc. Minor point bakes it more clear these are `bool` and also leaves these names available in case we need them later.

File: tidy3d/components/medium.py
Comment (tylerflex) [inline tidy3d/components/medium.py:690]: I think we can replace these 3 lines with just 
```py
return self.nonlinear or self.time_modulated or self.custom or self.fully_anisotropic
```

File: tidy3d/components/medium.py
Comment (tylerflex) [inline tidy3d/components/medium.py:685]: cached property? or maybe it's not such a big deal here.

File: tidy3d/components/simulation.py
Comment (tylerflex) [inline tidy3d/components/simulation.py:920]: let's put the code bits `(structures[{ind1}])` in single quotes `'` just to make the terminal output nicer.

File: tidy3d/components/simulation.py
Comment (tylerflex) [inline tidy3d/components/simulation.py:921]: let's insert the medium types while we're at it.
```py
... medium types of '{structure_ind1.medium.type}' and '{structure_ind2.medium.type}', respectively.
```

**Review by tylerflex — COMMENTED** [review]
much improved, thanks Casey! couple minor comments.

Also, I assume once #1478 is merged the tests might pass?

File: tidy3d/components/medium.py
Comment (tylerflex) [inline tidy3d/components/medium.py:699]: perhaps could be simplified / refactored a bit more

```py
@staticmethod
def _check_compatibility(med1, med2) -> bool:
    
    if self.custom:
        # custom and fully_anisotropic is OK
        if other.nonlinear or other.time_modulated:
            return False
    if self.fully_anisotropic:
        if other.nonlinear or other.time_modulated:
            return False
    if self.nonlinear:
        if other.time_modulated:
            return False    
    return True

def _compatible_with(self, other):
    if not (self._is_advanced and other._is_advanced):
        return True
    return self._check_compatibility(self, other) and self._check_compatibility(other, self)
```

File: tidy3d/components/medium.py
Comment (caseyflex) [reply tidy3d/components/medium.py:671]: I might leave time_modulated instead of changing it to is_time_modulated, just to avoid the need for a backend PR.

File: tidy3d/components/medium.py
Comment (caseyflex) [reply tidy3d/components/medium.py:686]: yeah, agreed that is nicer

**Comment by caseyflex** [comment]
> much improved, thanks Casey! couple minor comments.
> 
> Also, I assume once #1478 is merged the tests might pass?

I had to delete the json and h5 files for `2.5.1`. I wasn't quite sure if we are keeping those, or if so what the fix was to update SIM_FULL in this old file?

File: tidy3d/components/medium.py
Comment (tylerflex) [reply tidy3d/components/medium.py]: ok that's fine

**Comment by tylerflex** [comment]
> I had to delete the json and h5 files for 2.5.1. I wasn't quite sure if we are keeping those, or if so what the fix was to update SIM_FULL in this old file?

I'm not sure I understand, but after that PR, I think we are just maintaining a `simulation_sample.json` now (no explicit version numbers) and that file corresponds to the last minor version. I dont think there should have been a 2.5.1 so maybe that file got either generated by a test or wasn't deleted. Maybe @momchil-flex knows

File: tidy3d/components/medium.py
Comment (tylerflex) [reply tidy3d/components/medium.py]: hm, might be confusing if they follow different conventions?

**Comment by momchil-flex** [comment]
Hm yeah I don't know why 2.5.1 is still there but it can be deleted.

**Comment by caseyflex** [comment]
I renamed time_modulated to is_time_modulated and made a corresponding backend PR. Also, I added validators and tests to forbid nonlinear_spec or modulation_spec on components of a `Medium2D`.

**Review by weiliangjin2021 — APPROVED** [review]
If I understand correctly, the incompatibility is checked not against actual structure intersections, but bounding box? It sounds too strict, but fast to do.

**Comment by caseyflex** [comment]
> If I understand correctly, the incompatibility is checked not against actual structure intersections, but bounding box? It sounds too strict, but fast to do.

In GUI, if I understand correctly, they don't check intersections at all, just disallow conflicting types in the entire simulation.

**Comment by e-g-melo** [comment]
> > If I understand correctly, the incompatibility is checked not against actual structure intersections, but bounding box? It sounds too strict, but fast to do.
> 
> In GUI, if I understand correctly, they don't check intersections at all, just disallow conflicting types in the entire simulation.
> 
> I guess we can relax this in the future if required. But it would be better to make the advanced material properties compatible in the solver. This is meant to be temporary until we can support that functionality. I don't see a fast or easy way to actually check intersections on frontend. Do you?

That is correct. It was a temporary workaround in GUI, but indeed, we should allow the user to include different advanced material properties.

**Comment by tylerflex** [comment]
Looks good to me @caseyflex . Just a reminder about this [comment](https://github.com/flexcompute/tidy3d/pull/1474#discussion_r1492689266). I think we should change mentions of `advanced material` to mentions of `has compatibility`  in the code variables, tests, comments, changlog etc. Just for specificity and to avoid a future where we might need to introduce "really advanced materials" etc.  :) 

https://github.com/flexcompute/tidy3d/pull/1474#discussion_r1492689266

**Comment by caseyflex** [comment]
> Looks good to me @caseyflex . Just a reminder about this [comment](https://github.com/flexcompute/tidy3d/pull/1474#discussion_r1492689266). I think we should change mentions of `advanced material` to mentions of `has incompatibilities` in the code variables, tests, comments, changlog etc. Just for specificity and to avoid a future where we might need to introduce "really advanced materials" etc. :)
> 
> [#1474 (comment)](https://github.com/flexcompute/tidy3d/pull/1474#discussion_r1492689266)

done, should be fine to merge

**Comment by tylerflex** [comment]
Perhaps let's merge this in for 2.7? to avoid any potential issues? I'll leave it up to @momchil-flex

**Review by e-g-melo — APPROVED** [review]
Hi @caseyflex! Thanks for including this validation. I've just tried it using a simple setup and it worked fine.

I have only a comment, not related to this issue. Why do we show all these warnings after the validation error? 
image

**Comment by weiliangjin2021** [comment]
> > If I understand correctly, the incompatibility is checked not against actual structure intersections, but bounding box? It sounds too strict, but fast to do.
> 
> In GUI, if I understand correctly, they don't check intersections at all, just disallow conflicting types in the entire simulation.
> 
> I guess we can relax this in the future if required. But it would be better to make the advanced material properties compatible in the solver. This is meant to be temporary until we can support that functionality. I don't see a fast or easy way to actually check intersections on frontend. Do you?

I don't see an easy and efficient way either.

**Comment by momchil-flex** [comment]
> Hi @caseyflex! Thanks for including this validation. I've just tried it using a simple setup and it worked fine.
> 
> I have only a comment, not related to this issue. Why do we show all these warnings after the validation error? image

@tylerflex @lucas-flexcompute @dbochkov any idea if these warnings can be avoided e.g. with some pydantic settings?

**Comment by weiliangjin2021** [comment]
should this be merged to pre/2.7?

**Comment by weiliangjin2021** [comment]
> Is there a problem with it being in the next 2.6 patch?

In our previous workflow, the patch will be merged to develop. Has this been changed?

**Comment by momchil-flex** [comment]
Ok I see your point I thought this was merged in develop and you were just asking if it should go in 2.7 instead.

I'll fix tomorrow

**Comment by lucas-flexcompute** [comment]
> > Hi @caseyflex! Thanks for including this validation. I've just tried it using a simple setup and it worked fine.
> > I have only a comment, not related to this issue. Why do we show all these warnings after the validation error? image
> 
> @tylerflex @lucas-flexcompute @dbochkov any idea if these warnings can be avoided e.g. with some pydantic settings?

My understanding is that field validators are still executed if other validators have failed unless we set `skip_on_failure=True` on them, so that could be a solution.

**Pull request merged**
